logstash-core/src/main/java/org/logstash/plugins/factory/RubyCodecDelegator.java

```diff
@@ -0,0 +1,137 @@
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.Map;
+import java.util.function.Consumer;
+
+class RubyCodecDelegator implements Codec {
+
+    private static final Logger LOGGER = LogManager.getLogger(RubyCodecDelegator.class);
+
+    private final ThreadContext currentContext;
+    private final IRubyObject pluginInstance;
+    private final String wrappingId;
+
+    public RubyCodecDelegator(ThreadContext currentContext, IRubyObject pluginInstance) {
+        this.currentContext = currentContext;
+        this.pluginInstance = pluginInstance;
+
+        verifyCodecAncestry(pluginInstance);
+        invokeRubyRegister(currentContext, pluginInstance);
+
+        wrappingId = "jw-" + wrappedPluginId();
+    }
+
+    private String wrappedPluginId() {
+        RubyString id = (RubyString) pluginInstance.callMethod(this.currentContext, "id");
+        return id.toString();
+    }
+
+    private void verifyCodecAncestry(IRubyObject pluginInstance) {
+        final RubyClass codecBaseClass = RubyUtil.RUBY.getModule("LogStash").getModule("Codecs").getClass("Base");
+//            final boolean isRubyCodec = codecBaseClass.subclasses(true).contains(pluginInstance.getType());
+        final boolean isRubyCodec = pluginInstance.getType().hasModuleInHierarchy(codecBaseClass);
+        if (!isRubyCodec) {
+            throw new IllegalStateException("Ruby wrapped codec is expected to subclass LogStash::Codecs::Base");
+        }
+    }
+
+    private void invokeRubyRegister(ThreadContext currentContext, IRubyObject pluginInstance) {
+        pluginInstance.callMethod(currentContext, "register");
+    }
+
+    @Override
+    public void decode(ByteBuffer buffer, Consumer<Map<String, Object>> eventConsumer) {
+        // invoke Ruby's codec #decode(data, block) and use a Block to capture the yielded LogStash::Event to
+        // back to Java and pass to the eventConsumer.
+        if (buffer.remaining() == 0) {
+            // no data to decode
+            return;
+        }
+
+        // setup the block callback bridge to invoke eventConsumer
+        final Block consumerWrapper = new Block(new JavaInternalBlockBody(currentContext.runtime, Signature.ONE_ARGUMENT) {
+            @Override
+            @SuppressWarnings("unchecked")
+            public IRubyObject yield(ThreadContext context, IRubyObject[] args) {
+                // Expect only one argument, the LogStash::Event instantiated by the Ruby codec
+                final IRubyObject event = args[0];
+                // TODO should we check the type of the args[0] or assume it's always an RubyEvent?
+                final Map<String, Object> unwrappedEvent = ((JrubyEventExtLibrary.RubyEvent) event).ruby_to_hash(currentContext)
+                        .toJava(Map.class);
+                eventConsumer.accept(unwrappedEvent);
```

Comment:
is there a reason not simply unwrap the data stored on the Java event? e.g. 
```suggestion
                eventConsumer.accept( ((JrubyEventExtLibrary.RubyEvent) event).getEvent().getData() );
```

Reply:
Good point